Add NVLSM as dependency for NVIDIA Fabric Manager

yeazelm opened the pull request:
**Description of changes:**
New NVIDIA hardware now uses NVLSM (NVIDIA Switch Manager) to manage the nvswitches and Fabric Manager requires this tool to be running before it starts in this configuration. This adds the `nvlsm` package as its own package as well as the dependencies in Fabric Manager to ensure this service is running before it starts. This relies upon a new command line feature in `ghostdog` to detect this hardware and configure both Fabric Manager and NVLSM to use the same Port GUID for communication. These changes are in https://github.com/bottlerocket-os/bottlerocket-core-kit/pull/499.


**Testing done:**
Built an image with these changes and a core kit with the `ghostdog` changes. Before this change, Fabric Manager was unable to initialize properly. This left the GPUs in an uninitialized state. Now, they initialize correctly and workload tests pass.


Smoke tests pass:
<details>

```
# ./run.sh

=========================================
  Running sample UnifiedMemoryPerf
=========================================

GPU Device 0: "Blackwell" with compute capability 10.0

Running ........................................................

Overall Time For matrixMultiplyPerf

Printing Average of 20 measurements in (ms)
Size_KB  UMhint UMhntAs  UMeasy   0Copy MemCopy CpAsync CpHpglk CpPglAs
4         0.310   0.381   0.498   0.017   0.038   0.028   0.043   0.028
16        0.338   0.463   0.734   0.031   0.058   0.050   0.056   0.042
64        0.533   0.815   0.894   0.065   0.119   0.114   0.089   0.078
256       0.824   0.857   1.488   0.234   0.244   0.228   0.208   0.195
1024      1.644   1.891   2.240   1.203   0.878   0.849   0.700   0.722
4096      4.412   4.303   7.628   7.220   3.300   3.320   2.714   2.751
16384    17.734  16.525  26.929  42.127  13.531  13.546  11.289  11.418

NOTE: The CUDA Samples are not meant for performance measurements. Results may vary when GPU Boost is enabled.

=========================================
  Running sample deviceQuery
=========================================

./deviceQuery Starting...

 CUDA Device Query (Runtime API) version (CUDART static linking)

Detected 1 CUDA Capable device(s)

Device 0: "NVIDIA B200"
  CUDA Driver Version / Runtime Version          12.8 / 12.8
  CUDA Capability Major/Minor version number:    10.0
  Total amount of global memory:                 182642 MBytes (191514411008 bytes)
  (148) Multiprocessors, (128) CUDA Cores/MP:    18944 CUDA Cores
  GPU Max Clock rate:                            1965 MHz (1.97 GHz)
  Memory Clock rate:                             3996 Mhz
  Memory Bus Width:                              7680-bit
  L2 Cache Size:                                 132644864 bytes
  Maximum Texture Dimension Size (x,y,z)         1D=(131072), 2D=(131072, 65536), 3D=(16384, 16384, 16384)
  Maximum Layered 1D Texture Size, (num) layers  1D=(32768), 2048 layers
  Maximum Layered 2D Texture Size, (num) layers  2D=(32768, 32768), 2048 layers
  Total amount of constant memory:               65536 bytes
  Total amount of shared memory per block:       49152 bytes
  Total shared memory per multiprocessor:        233472 bytes
  Total number of registers available per block: 65536
  Warp size:                                     32
  Maximum number of threads per multiprocessor:  2048
  Maximum number of threads per block:           1024
  Max dimension size of a thread block (x,y,z): (1024, 1024, 64)
  Max dimension size of a grid size    (x,y,z): (2147483647, 65535, 65535)
  Maximum memory pitch:                          2147483647 bytes
  Texture alignment:                             512 bytes
  Concurrent copy and kernel execution:          Yes with 4 copy engine(s)
  Run time limit on kernels:                     No
  Integrated GPU sharing Host Memory:            No
  Support host page-locked memory mapping:       Yes
  Alignment requirement for Surfaces:            Yes
  Device has ECC support:                        Enabled
  Device supports Unified Addressing (UVA):      Yes
  Device supports Managed Memory:                Yes
  Device supports Compute Preemption:            Yes
  Supports Cooperative Kernel Launch:            Yes
  Supports MultiDevice Co-op Kernel Launch:      Yes
  Device PCI Domain ID / Bus ID / location ID:   0 / 81 / 0
  Compute Mode:
     < Default (multiple host threads can use ::cudaSetDevice() with device simultaneously) >

deviceQuery, CUDA Driver = CUDART, CUDA Driver Version = 12.8, CUDA Runtime Version = 12.8, NumDevs = 1
Result = PASS
....
```

</details>

The env file at `/etc/nvidia/nvlsm.env` has the guid args variable in it:

<details>

```
bash-5.1# cat /etc/nvidia/nvlsm.env
GUID_ARG="-g 0xe09d7303003f3bf8"
```

</details>

New boot has nvlsm in the critical chain (the 5 second sleep is in the nvlsm-online service):
```
bash-5.1# systemd-analyze
Startup finished in 792ms (firmware) + 913ms (loader) + 3.909s (kernel) + 41.341s (userspace) = 46.955s
multi-user.target reached after 41.292s in userspace.
bash-5.1# systemd-analyze critical-chain
The time when unit became active or started is printed after the "@" character.
The time the unit took to start is printed after the "+" character.

multi-user.target @41.292s
└─nvidia-fabricmanager.service @41.291s
  └─nvlsm-online.service @36.171s +5.005s
    └─nvlsm.service @36.163s +3ms
      └─infiniband-guid.service @36.137s +15ms
        └─basic.target @12.953s
          └─dbus-broker.service @12.926s +24ms
            └─dbus.socket @12.923s
              └─sysinit.target @12.920s
                └─modprobe@ib_umad.service @36.141s +22ms
                  └─systemd-journald.socket @1.106s
                    └─system.slice @1.058s
                      └─-.slice @1.058s

```

Fabric manger and nvlsm services are healthy:

<details>

```
bash-5.1# systemctl status nvidia-fabricmanager
● nvidia-fabricmanager.service - NVIDIA fabric manager service
     Loaded: loaded (/x86_64-bottlerocket-linux-gnu/sys-root/usr/lib/systemd/system/nvidia-fabricmanager.service; enabled; preset: enabled)
    Drop-In: /x86_64-bottlerocket-linux-gnu/sys-root/usr/lib/systemd/system/service.d
             └─00-aws-config.conf
     Active: active (running) since Fri 2025-05-16 15:10:10 UTC; 13min ago
   Main PID: 29850 (nv-fabricmanage)
      Tasks: 32 (limit: 629145)
     Memory: 20.2M
        CPU: 1.861s
     CGroup: /system.slice/nvidia-fabricmanager.service
             └─29850 /usr/bin/nv-fabricmanager -c /etc/nvidia/fabricmanager.cfg -g 0xe09d7303003f3bf8

May 16 15:10:22 ip-192-168-38-187.us-west-2.compute.internal nv-fabricmanager[29850]: [May 16 2025 15:10:21] [INFO] [tid 30042] GPU probe request received from node Id 0 and module Id 7 GUID 0x3ba54f661add0c35 on the partition 57082 requestId 0x18400a7ecabe8408.
May 16 15:10:22 ip-192-168-38-187.us-west-2.compute.internal nv-fabricmanager[29850]: [May 16 2025 15:10:21] [INFO] [tid 30042] received GPU Probe Request ver 2 MAD: header nodeID 0, header moduleId 7, status 0, nodeId 0, gpuGuid 0x3ba54f661add0c35, requestId 0x18400a7eca…
May 16 15:10:22 ip-192-168-38-187.us-west-2.compute.internal nv-fabricmanager[29850]: [May 16 2025 15:10:21] [INFO] [tid 30042] sent GPU Probe Response ver 2 MAD: header nodeID 0, header moduleId 7, status 0, nodeId 0, gpuGuid 0x3ba54f661add0c35, requestId 0x18400a7ecabe8…
May 16 15:10:22 ip-192-168-38-187.us-west-2.compute.internal nv-fabricmanager[29850]: [May 16 2025 15:10:21] [INFO] [tid 30042] GPU probe request received from node Id 0 and module Id 0 GUID 0x46ec5d0ac7813c0c on the partition 57082 requestId 0x18400a7dabcd5778.
May 16 15:10:22 ip-192-168-38-187.us-west-2.compute.internal nv-fabricmanager[29850]: [May 16 2025 15:10:21] [INFO] [tid 30042] received GPU Probe Request ver 2 MAD: header nodeID 0, header moduleId 0, status 0, nodeId 0, gpuGuid 0x46ec5d0ac7813c0c, requestId 0x18400a7dab…
May 16 15:10:22 ip-192-168-38-187.us-west-2.compute.internal nv-fabricmanager[29850]: [May 16 2025 15:10:21] [INFO] [tid 30042] sent GPU Probe Response ver 2 MAD: header nodeID 0, header moduleId 0, status 0, nodeId 0, gpuGuid 0x46ec5d0ac7813c0c, requestId 0x18400a7dabcd5…
May 16 15:10:22 ip-192-168-38-187.us-west-2.compute.internal nv-fabricmanager[29850]: [May 16 2025 15:10:21] [INFO] [tid 30042] GPU probe request received from node Id 0 and module Id 3 GUID 0xc023b36e716620b0 on the partition 57082 requestId 0x18400a7c8a9f0698.
May 16 15:10:22 ip-192-168-38-187.us-west-2.compute.internal nv-fabricmanager[29850]: [May 16 2025 15:10:21] [INFO] [tid 30042] received GPU Probe Request ver 2 MAD: header nodeID 0, header moduleId 3, status 0, nodeId 0, gpuGuid 0xc023b36e716620b0, requestId 0x18400a7c8a…
May 16 15:10:22 ip-192-168-38-187.us-west-2.compute.internal nv-fabricmanager[29850]: [May 16 2025 15:10:21] [INFO] [tid 30042] sent GPU Probe Response ver 2 MAD: header nodeID 0, header moduleId 3, status 0, nodeId 0, gpuGuid 0xc023b36e716620b0, requestId 0x18400a7c8a9f0…
May 16 15:10:23 ip-192-168-38-187.us-west-2.compute.internal nv-fabricmanager[29850]: [May 16 2025 15:10:23] [INFO] [tid 30176] fabric management state monitor detected fabric manager state update; fabric manager state is `3` (configured)


bash-5.1# systemctl status nvlsm
● nvlsm.service - Start NVLink Subnet Manager
     Loaded: loaded (/x86_64-bottlerocket-linux-gnu/sys-root/usr/lib/systemd/system/nvlsm.service; enabled; preset: enabled)
    Drop-In: /x86_64-bottlerocket-linux-gnu/sys-root/usr/lib/systemd/system/service.d
             └─00-aws-config.conf
     Active: active (running) since Fri 2025-05-16 15:10:04 UTC; 14min ago
   Main PID: 28974 (nvlsm)
      Tasks: 58 (limit: 629145)
     Memory: 43.3M
        CPU: 1.678s
     CGroup: /system.slice/nvlsm.service
             ├─28974 /usr/bin/nvlsm -F /usr/share/nvidia/nvlsm/nvlsm.conf -f stdout --pid_file /var/run/nvidia-fabricmanager/nvlsm.pid -g 0xe09d7303003f3bf8
             └─29072 osm_crashd

May 16 15:24:04 ip-192-168-38-187.us-west-2.compute.internal nvlsm[28974]: May 16 15:24:04 118453 [3F3EF6C0] 0x02 -> do_sweep:
May 16 15:24:04 ip-192-168-38-187.us-west-2.compute.internal nvlsm[28974]: ******************************************************************
May 16 15:24:04 ip-192-168-38-187.us-west-2.compute.internal nvlsm[28974]: *********************** HEAVY SWEEP START ************************
May 16 15:24:04 ip-192-168-38-187.us-west-2.compute.internal nvlsm[28974]: ******************************************************************
May 16 15:24:04 ip-192-168-38-187.us-west-2.compute.internal nvlsm[28974]: May 16 15:24:04 118460 [3F3EF6C0] 0x02 -> do_sweep: Entering heavy sweep with flags: force_heavy_sweep 1, coming out of standby 0, subnet initialization error 0, sm port change 0
May 16 15:24:04 ip-192-168-38-187.us-west-2.compute.internal nvlsm[28974]: AR Manager - Configuration cycle (number 58) completed successfully
May 16 15:24:04 ip-192-168-38-187.us-west-2.compute.internal nvlsm[28974]: May 16 15:24:04 131605 [3F3EF6C0] 0x02 -> osm_ucast_mgr_process: ar_minhop tables configured on all switches
May 16 15:24:04 ip-192-168-38-187.us-west-2.compute.internal nvlsm[28974]: May 16 15:24:04 131645 [3F3EF6C0] 0x02 -> osm_ar_calculate_pfrn: No fabric switch supports pFRN. Hence, avoid configuring pFRN
May 16 15:24:04 ip-192-168-38-187.us-west-2.compute.internal nvlsm[28974]: May 16 15:24:04 136153 [3F3EF6C0] 0x02 -> SUBNET UP
May 16 15:24:04 ip-192-168-38-187.us-west-2.compute.internal nvlsm[28974]: May 16 15:24:04 136821 [3F3EF6C0] 0x02 -> GFM_PLUGIN - HandleEvents: Pkey sweep done
```

</details>


Also validated on an existing `p4d.24xlarge` that Fabric Manager starts and workloads work on it still:

Sample workload test: 

<details>

```
cat << EOF | kubectl apply -f -
apiVersion: batch/v1
kind: Job
metadata:
  name: nvidia-cuda-workload
spec:
  parallelism: 20
  completions: 20
  template:
    spec:
      containers:
      - name: vector-add
        image: nvcr.io/nvidia/k8s/cuda-sample:vectoradd-cuda12.5.0-ubi8
        resources:
          limits:
            nvidia.com/gpu: 1
      restartPolicy: Never
EOF

$ kubectl logs nvidia-cuda-workload-5fs2k
[Vector addition of 50000 elements]
Copy input data from the host memory to the CUDA device
CUDA kernel launch with 196 blocks of 256 threads
Copy output data from the CUDA device to the host memory
Test PASSED
Done
```

</details>

From the nvidia-fabricmanager logs: 

<details>

```
May 12 15:08:04 ip-192-168-92-236.us-west-2.compute.internal nv-fabricmanager[15992]: [May 12 2025 15:08:04] [INFO] [tid 15992] start NVSwitch 0/12 routing configuration
May 12 15:08:04 ip-192-168-92-236.us-west-2.compute.internal nv-fabricmanager[15992]: [May 12 2025 15:08:04] [INFO] [tid 15992] completed NVSwitch 0/12 routing configuration
May 12 15:08:04 ip-192-168-92-236.us-west-2.compute.internal nv-fabricmanager[15992]: [May 12 2025 15:08:04] [INFO] [tid 15992] start NVSwitch 0/13 routing configuration
May 12 15:08:04 ip-192-168-92-236.us-west-2.compute.internal nv-fabricmanager[15992]: [May 12 2025 15:08:04] [INFO] [tid 15992] completed NVSwitch 0/13 routing configuration
May 12 15:08:04 ip-192-168-92-236.us-west-2.compute.internal nv-fabricmanager[15992]: [May 12 2025 15:08:04] [INFO] [tid 15992] GPU with nodeId = 0, gpuId = 1 is configured
May 12 15:08:04 ip-192-168-92-236.us-west-2.compute.internal nv-fabricmanager[15992]: [May 12 2025 15:08:04] [INFO] [tid 15992] GPU with nodeId = 0, gpuId = 2 is configured
May 12 15:08:04 ip-192-168-92-236.us-west-2.compute.internal nv-fabricmanager[15992]: [May 12 2025 15:08:04] [INFO] [tid 15992] GPU with nodeId = 0, gpuId = 3 is configured
May 12 15:08:04 ip-192-168-92-236.us-west-2.compute.internal nv-fabricmanager[15992]: [May 12 2025 15:08:04] [INFO] [tid 15992] GPU with nodeId = 0, gpuId = 4 is configured
May 12 15:08:04 ip-192-168-92-236.us-west-2.compute.internal nv-fabricmanager[15992]: [May 12 2025 15:08:04] [INFO] [tid 15992] GPU with nodeId = 0, gpuId = 5 is configured
May 12 15:08:04 ip-192-168-92-236.us-west-2.compute.internal nv-fabricmanager[15992]: [May 12 2025 15:08:04] [INFO] [tid 15992] GPU with nodeId = 0, gpuId = 6 is configured
May 12 15:08:04 ip-192-168-92-236.us-west-2.compute.internal nv-fabricmanager[15992]: [May 12 2025 15:08:04] [INFO] [tid 15992] GPU with nodeId = 0, gpuId = 7 is configured
May 12 15:08:04 ip-192-168-92-236.us-west-2.compute.internal nv-fabricmanager[15992]: [May 12 2025 15:08:04] [INFO] [tid 15992] GPU with nodeId = 0, gpuId = 8 is configured
May 12 15:08:04 ip-192-168-92-236.us-west-2.compute.internal nv-fabricmanager[15992]: [May 12 2025 15:08:04] [INFO] [tid 15992] Multi node HA error handling disabled
May 12 15:08:04 ip-192-168-92-236.us-west-2.compute.internal nv-fabricmanager[15992]: [May 12 2025 15:08:04] [INFO] [tid 15992] Successfully configured all the available GPUs and NVSwitches to route NVLink traffic.
```

</details>

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

Commit:
packages/nvlsm: Add in nvlsm package

This package is needed for NVIDIA Fabric Manager. It is used for
managing nvswitches. This adds a service that depends on a call to
ghostdog to check if hardware is present so it only runs if required.

Co-authored-by: Patrick J.P. Culp <jpculp@amazon.com>
Signed-off-by: Matthew Yeazel <yeazelm@amazon.com>

Commit:
kmod-6.1-nvidia-r570: Add dependency on nvlsm

Fabric Manager now needs NVLSM for some hardware configurations. This
adds that dependency to the rpm and nvidia-fabricmanager service so that
nvlsm is up and running before Fabric Manager attempts to initalize.

Signed-off-by: Matthew Yeazel <yeazelm@amazon.com>

Commit:
kmod-6.12-nvidia-r570: Add dependency on nvlsm

Fabric Manager now needs NVLSM for some hardware configurations. This
adds that dependency to the rpm and nvidia-fabricmanager service so that
nvlsm is up and running before Fabric Manager attempts to initalize.

Signed-off-by: Matthew Yeazel <yeazelm@amazon.com>

yeazelm commented:
^ Adding in additional author to nvlsm commit.

bcressey commented on packages/nvlsm/nvlsm.spec:
```diff
@@ -0,0 +1,92 @@
+Release: 1%{?dist}
+Epoch: 1
+Summary: NVIDIA Subnet Manager component
+License: LicenseRef-NVIDIA-AWS-EULA
```
This license statement doesn't seem correct.

yeazelm replied:
I was able to verify this license

bcressey commented on packages/nvlsm/nvlsm.spec:
```diff
@@ -0,0 +1,92 @@
+Name: %{_cross_os}nvlsm
+Version: %{nvlsm_ver}
+Release: 1%{?dist}
+Epoch: 1
```
Why bump the epoch for a new package?

yeazelm replied:
No good reason, just was brought over from the other packages. I'll remove.

bcressey commented on packages/nvlsm/nvlsm.spec:
```diff
@@ -0,0 +1,92 @@
+%global nvlsm_minor 03
+%global nvlsm_patch 1
+%global nvlsm_ver %{nvlsm_major}.%{nvlsm_minor}.%{nvlsm_patch}
+%global __brp_check_rpaths %{nil}
```
Need to add a comment explaining this.

bcressey commented on packages/nvlsm/nvlsm.spec:
```diff
@@ -0,0 +1,92 @@
+Epoch: 1
+Summary: NVIDIA Subnet Manager component
+License: LicenseRef-NVIDIA-AWS-EULA
+URL: http://nvidia.com
```
```suggestion
URL: https://nvidia.com
```

jpculp replied:
I could go either way, but the `http` is consistent with the upstream packages and the kmods.

bcressey commented on packages/nvlsm/nvlsm.spec:
```diff
@@ -0,0 +1,92 @@
+# NVIDIA rpms from 0 to 199
+Source0: https://developer.download.nvidia.com/compute/cuda/repos/rhel9/x86_64/nvlsm-%{nvlsm_ver}-1.x86_64.rpm
+Source1: https://developer.download.nvidia.com/compute/cuda/repos/rhel9/sbsa/nvlsm-%{nvlsm_ver}-1.aarch64.rpm
+Source2: NVidiaEULAforAWS.pdf
```
This is not an NVIDIA rpm and it's in the sources from 0 to 199.

yeazelm replied:
This convention is consistent with the kmod packages, I can change it but we should change it for the others as well.

bcressey commented on packages/nvlsm/nvlsm.spec:
```diff
@@ -0,0 +1,92 @@
+install -d %{buildroot}%{_cross_factorydir}%{_cross_sysconfdir}/nvidia
+install -p -m 0644 usr/share/nvidia/nvlsm/nvlsm.conf %{buildroot}%{_cross_factorydir}%{_cross_sysconfdir}/nvidia/nvlsm.conf
+
+# Configuration files that are referenced but do not need to be updated at runtime
+install -d %{buildroot}%{_cross_datadir}/nvidia/nvlsm
+install -p -m 0644 usr/share/nvidia/nvlsm/device_configuration.conf %{buildroot}%{_cross_datadir}/nvidia/nvlsm
+install -p -m 0644 usr/share/nvidia/nvlsm/grpc_mgr.conf %{buildroot}%{_cross_datadir}/nvidia/nvlsm
```
Seems complicated to put config files in two different places - is this necessary?

yeazelm replied:
This is because we need to modify the primary config to add in a hardware GUID but the other two files are actually better left read only. We can move them all to the factory but I opted for the slightly more complicated directory structure to retain a bit of durability against accidental change.

yeazelm replied:
I found a way to pass `-g` instead of editing the config file. This simplifies a lot of things and one of them is the need to edit this file. I'll keep everything in the `%{_cross_datadir}`.

bcressey commented on packages/nvlsm/nvlsm.service:
```diff
@@ -0,0 +1,15 @@
+RequiresMountsFor=/etc
+
+[Service]
+Type=oneshot
```
Is this really a one-shot?

yeazelm replied:
This is probably just an `exec` rather than a `oneshot`. If we avoid using `-B` for daemonizing, I think that would give us a `simple` instead but I'd have to experiment with this to see if that is the behavior we really want.

bcressey commented on packages/nvlsm/nvlsm.service:
```diff
@@ -0,0 +1,15 @@
+[Service]
+Type=oneshot
+ExecCondition=/usr/bin/ghostdog infiniband-present
+ExecStart=/usr/bin/nvlsm -F /etc/nvidia/nvlsm.conf -B --pid_file /var/run/nvidia-fabricmanager/nvlsm.pid
```
Can this be made to run in the foreground, like Fabric Manager, rather than letting it do its own pid file based service management?

yeazelm replied:
I can take a look, I used the default behavior provided by the Fabric Manager start script that uses this but I think  we should consider what flags are ideal for our use. I'll see if I can get nvlsm to write to stdout.

yeazelm replied:
I updated the service to run in the forerground and output logs to standard out.

bcressey commented on packages/nvlsm/nvlsm.service:
```diff
@@ -0,0 +1,15 @@
+ExecCondition=/usr/bin/ghostdog infiniband-present
+ExecStart=/usr/bin/nvlsm -F /etc/nvidia/nvlsm.conf -B --pid_file /var/run/nvidia-fabricmanager/nvlsm.pid
+# nvlsm requires time to start
+ExecStartPost=/bin/sleep 5
```
The `sleep 5` is a giant hack, is there any external indicator that the service is ready that can be used instead?

yeazelm replied:
From what I could tell, the files are created immediately, but the process won't actually respond to the gRPC calls from Fabric Manager resulting in a bad state from Fabric Manager. This logic consistent with the NVIDIA `nvidia-fabricmanager-start.sh` provided by the Fabric Manager rpm. I can look for more places to check but I wasn't able to find much in my initial search for a better way than a default `sleep 5`.

yeazelm replied:
I moved this to a separate service file. I have not yet found a clear signal that would remove the sleep but it only occurs when nvlsm has been started so the delay only happens for this particular situation where the system needs more time.

bcressey commented on packages/kmod-6.12-nvidia-r570/nvidia-fabricmanager.service:
```diff
@@ -2,6 +2,8 @@
 Description=NVIDIA fabric manager service
 Wants=modprobe@ib_umad.service
 After=modprobe@ib_umad.service
+Requires=nvlsm.service
+After=nvlsm.service
```
This "after" dependency should go on the same line as the modprobe one.

bcressey commented on packages/nvlsm/Cargo.toml:
```diff
@@ -0,0 +1,21 @@
+[lib]
+path = "../packages.rs"
+
+[[package.metadata.build-package.external-files]]
```
Don't we need `NVidiaEULAforAWS.pdf` as another external file?

yeazelm replied:
Yes, I'll fix in the next revision.

arnaldo2792 commented on packages/nvlsm/nvlsm.spec:
```diff
@@ -0,0 +1,92 @@
+install -p -m 0644 %{S:201} %{buildroot}%{_cross_tmpfilesdir}/nvlsm.conf
+
+install -d %{buildroot}%{_cross_factorydir}%{_cross_sysconfdir}/ld.so.conf.d/
+# We need to add `_cross_libdir` to the paths loaded by the ldconfig service
```
nit: we should update all other comments, but we don't just add `%{_cross_libdir}`, what's added is `%{_cross_libdir}/nvidia/nvlsm`.

yeazelm replied:
The .in files include `/nvidia/nvlsm` already but I can update the comment to reflect its actually `%{_cross_libdir}` being replaced in this file.

yeazelm commented:
^ Addresses feedback on the systemd services and nvlsm, but not the rdma-core changes. I'll work on getting those included but this matches the new approach in https://github.com/bottlerocket-os/bottlerocket-core-kit/pull/499 so both reviews can be coordinated.

yeazelm commented:
^ This push updates the systemd dependencies to work on non-fabric manager instances (such as g5.xlarge) as well as non B200 Fabric Manager instances like p4d.24xlarge. Validated that the units for nvlsm and nvlsm-online are skipped when infiniband devices are not present.

bcressey commented on packages/nvlsm/infiniband-guid.service:
```diff
@@ -0,0 +1,13 @@
+
```
```suggestion
```

bcressey commented on packages/nvlsm/nvlsm-online.service:
```diff
@@ -0,0 +1,14 @@
+[Unit]
+Description=Start NVLink Subnet Manager Online Service
```
```suggestion
Description=Wait for NVLink Subnet Manager
```

bcressey commented on packages/nvlsm/infiniband-guid.service:
```diff
@@ -0,0 +1,13 @@
+
+[Unit]
+Description=Detect Infiniband devices and write the GUID env file if present
```
```suggestion
Description=Write GUID for NVLink Subnet Manager
```

bcressey commented on packages/nvlsm/infiniband-guid.service:
```diff
@@ -0,0 +1,13 @@
+
+[Unit]
+Description=Detect Infiniband devices and write the GUID env file if present
+RequiresMountsFor=/etc
```
You shouldn't need these `RequiresMountsFor` rules for a unit with default dependencies.
```suggestion
```

bcressey commented on packages/nvlsm/nvlsm.service:
```diff
@@ -0,0 +1,17 @@
+[Unit]
+Description=Start NVLink Subnet Manager
```
```suggestion
Description=NVLink Subnet Manager
```

bcressey commented on packages/nvlsm/nvlsm.service:
```diff
@@ -0,0 +1,17 @@
+[Unit]
+Description=Start NVLink Subnet Manager
+RequiresMountsFor=/etc
```
```suggestion
```

yeazelm replied:
I rebuilt and tested without this change and the behavior looks the same.

yeazelm commented:
^ Updated with comments to improve systemd unit naming and remove the unneeded /etc/ dependency.

<details>

```
systemctl status
● ip-192-168-62-187.us-west-2.compute.internal
    State: running
    Units: 534 loaded (incl. loaded aliases)
     Jobs: 0 queued
   Failed: 0 units
    Since: Mon 2025-05-19 23:48:21 UTC; 3min 43s ago
  systemd: 252
   CGroup: /
....

bash-5.1# systemctl status nvlsm
● nvlsm.service - NVLink Subnet Manager
     Loaded: loaded (/x86_64-bottlerocket-linux-gnu/sys-root/usr/lib/systemd/system/nvlsm.service; enabled; preset: enabled)
    Drop-In: /x86_64-bottlerocket-linux-gnu/sys-root/usr/lib/systemd/system/service.d
             └─00-aws-config.conf
     Active: active (running) since Mon 2025-05-19 23:49:00 UTC; 5min ago
   Main PID: 29031 (nvlsm)
      Tasks: 58 (limit: 629145)
     Memory: 43.7M
        CPU: 847ms
     CGroup: /system.slice/nvlsm.service
             ├─29031 /usr/bin/nvlsm -F /usr/share/nvidia/nvlsm/nvlsm.conf -f stdout --pid_file /var/run/nvidia-fabricmanager/nvlsm.pid -g 0xe09d7303003c08e2
             └─29084 osm_crashd

bash-5.1# systemctl status nvlsm-online
● nvlsm-online.service - Wait for NVLink Subnet Manager
     Loaded: loaded (/x86_64-bottlerocket-linux-gnu/sys-root/usr/lib/systemd/system/nvlsm-online.service; enabled; preset: enabled)
    Drop-In: /x86_64-bottlerocket-linux-gnu/sys-root/usr/lib/systemd/system/service.d
             └─00-aws-config.conf
     Active: active (exited) since Mon 2025-05-19 23:49:06 UTC; 6min ago
   Main PID: 29070 (code=exited, status=0/SUCCESS)
        CPU: 3ms

May 19 23:49:00 ip-192-168-62-187.us-west-2.compute.internal systemd[1]: Starting Wait for NVLink Subnet Manager...
May 19 23:49:06 ip-192-168-62-187.us-west-2.compute.internal systemd[1]: Finished Wait for NVLink Subnet Manager.

bash-5.1# systemctl status infiniband-guid
● infiniband-guid.service - Write GUID for NVLink Subnet Manager
     Loaded: loaded (/x86_64-bottlerocket-linux-gnu/sys-root/usr/lib/systemd/system/infiniband-guid.service; enabled; preset: enabled)
    Drop-In: /x86_64-bottlerocket-linux-gnu/sys-root/usr/lib/systemd/system/service.d
             └─00-aws-config.conf
     Active: active (exited) since Mon 2025-05-19 23:49:00 UTC; 6min ago
   Main PID: 28845 (code=exited, status=0/SUCCESS)
        CPU: 8ms

May 19 23:49:00 ip-192-168-62-187.us-west-2.compute.internal systemd[1]: Starting Write GUID for NVLink Subnet Manager...
May 19 23:49:00 ip-192-168-62-187.us-west-2.compute.internal systemd[1]: Finished Write GUID for NVLink Subnet Manager.
```

</details>

yeazelm commented:
^ One more improvement to remove `RemainAfterExit` and replace it with `Restart=Always` which is the correct behavior now that this is running in the foreground.
